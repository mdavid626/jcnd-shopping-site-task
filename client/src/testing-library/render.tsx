/* istanbul ignore file */
import {
  ApolloClient,
  ApolloProvider,
  InMemoryCache,
  NormalizedCacheObject,
} from '@apollo/client';
import { InitialEntry, Location } from '@remix-run/router';
import { Queries, queries } from '@testing-library/react';
import {
  RenderHookOptions,
  RenderOptions,
  render,
  renderHook,
  RenderResult,
  RenderHookResult,
} from '@testing-library/react';
import React, { ReactElement, useContext } from 'react';
import { MemoryRouter, useLocation, useNavigate } from 'react-router-dom';
import { NavigateFunction } from 'react-router/dist/lib/hooks';
import ShoppingCartContext from '../contexts/shopping-cart-context/shopping-cart-context';
import ShoppingCartProvider from '../contexts/shopping-cart-context/shopping-cart-provider';
import {
  ShoppingCartContextValue,
  ShoppingCartItem,
} from '../types/shopping-cart';

type SimpleRouter = {
  location?: Location;
  navigate?: NavigateFunction;
};

type RouterResult = {
  router: SimpleRouter;
};

type ApolloClientResult = {
  apolloClient: ApolloClient<NormalizedCacheObject>;
};

type ShoppingCartContextResult = {
  shoppingCartContext: { value?: ShoppingCartContextValue };
};

export const renderWithRouter = (
  component: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialEntries?: InitialEntry[]
): RenderResult & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...render(component, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  } as RenderResult & RouterResult;
};

export const renderWithFullContext = (
  component: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialEntries?: InitialEntry[]
): RenderResult &
  RouterResult &
  ApolloClientResult &
  ShoppingCartContextResult => {
  const router = {} as SimpleRouter;
  const apolloClient = new ApolloClient({
    uri: 'https://api.github.com/graphql',
    cache: new InMemoryCache(),
  });
  const shoppingCartContext: { value?: ShoppingCartContextValue } = {};
  return {
    ...render(component, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          shoppingCartContext.value = useContext(ShoppingCartContext);
          return children;
        };
        return (
          <>
            <ApolloProvider client={apolloClient}>
              <ShoppingCartProvider>
                <MemoryRouter initialEntries={initialEntries}>
                  <Children />
                </MemoryRouter>
              </ShoppingCartProvider>
            </ApolloProvider>
            <div id="modal-root" data-testid="modal-root" />
          </>
        );
      },
      ...options,
    }),
    router,
    apolloClient,
    shoppingCartContext,
  } as RenderResult &
    RouterResult &
    ApolloClientResult &
    ShoppingCartContextResult;
};

export const renderHookWithRouter = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>,
  initialEntries?: InitialEntry[]
): RenderHookResult<Result, Props> & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  };
};

export const renderHookWithQueryClient = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>
): RenderHookResult<Result, Props> & ApolloClientResult => {
  const apolloClient = new ApolloClient({
    uri: 'https://api.github.com/graphql',
    cache: new InMemoryCache(),
  });
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => (
        <ApolloProvider client={apolloClient}>{children}</ApolloProvider>
      ),
      ...options,
    }),
    apolloClient,
  };
};

export const renderHookWithShoppingCart = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>,
  initialShoppingCart?: ShoppingCartItem[]
): RenderHookResult<Result, Props> & ShoppingCartContextResult => {
  let shoppingCartContext: { value?: ShoppingCartContextValue } = {};
  window.sessionStorage.setItem(
    'shoppingCart',
    JSON.stringify(initialShoppingCart || [])
  );
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => {
        const Children = () => {
          shoppingCartContext.value = useContext(ShoppingCartContext);
          return children;
        };
        return (
          <ShoppingCartProvider>
            <Children />
          </ShoppingCartProvider>
        );
      },
      ...options,
    }),
    shoppingCartContext,
  };
};

export const renderHookWithShoppingCartAndQueryClientAndRouter = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>,
  initialShoppingCart?: ShoppingCartItem[],
  initialEntries?: string[]
): RenderHookResult<Result, Props> &
  ShoppingCartContextResult &
  ApolloClientResult &
  RouterResult => {
  const router = {} as SimpleRouter;
  const apolloClient = new ApolloClient({
    uri: 'https://api.github.com/graphql',
    cache: new InMemoryCache(),
  });
  window.sessionStorage.setItem(
    'shoppingCart',
    JSON.stringify(initialShoppingCart || [])
  );
  let shoppingCartContext: { value?: ShoppingCartContextValue } = {};
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          shoppingCartContext.value = useContext(ShoppingCartContext);
          return children;
        };
        return (
          <ApolloProvider client={apolloClient}>
            <ShoppingCartProvider>
              <MemoryRouter initialEntries={initialEntries}>
                <Children />
              </MemoryRouter>
            </ShoppingCartProvider>
          </ApolloProvider>
        );
      },
      ...options,
    }),
    shoppingCartContext,
    apolloClient,
    router,
  };
};
