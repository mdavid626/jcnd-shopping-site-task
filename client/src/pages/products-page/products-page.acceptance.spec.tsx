import { cleanup, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import fetchMock from 'jest-fetch-mock';
import React from 'react';
import Routes from '../../router/routes';
import {
  placeOrderQueryResult1,
  productsQueryResult,
} from '../../test-data/query-results';
import { renderWithFullContext } from '../../testing-library/render';

describe('[Acceptance] products-page', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    fetchMock.mockResponseOnce(JSON.stringify(productsQueryResult));
    fetchMock.mockResponseOnce(JSON.stringify(placeOrderQueryResult1));
  });
  afterEach(cleanup);
  afterEach(jest.resetAllMocks);

  it('should be able see products', async () => {
    renderWithFullContext(<Routes />);

    await screen.findByText('Broccoli');
    const product1 = screen.getByTestId('ProductItem-63ea6485732ff5cf8b364431');
    expect(within(product1).getByText('In stock: 466')).toBeVisible();

    expect(screen.getByText('Cabbage')).toBeVisible();
    const product2 = screen.getByTestId('ProductItem-63ea6485732ff5cf8b364434');
    expect(within(product2).getByText('In stock: 200')).toBeVisible();
  });

  it('should be able to buy product', async () => {
    renderWithFullContext(<Routes />);
    await screen.findByText('Broccoli');

    const product1 = screen.getByTestId('ProductItem-63ea6485732ff5cf8b364431');
    expect(within(product1).getByText('In stock: 466')).toBeVisible();
    await userEvent.click(within(product1).getByText('Buy'));
    expect(within(product1).getByText('In stock: 465')).toBeVisible();

    expect(screen.getByText('Cabbage')).toBeVisible();
    const product2 = screen.getByTestId('ProductItem-63ea6485732ff5cf8b364434');
    expect(within(product2).getByText('In stock: 200')).toBeVisible();
    await userEvent.click(within(product2).getByText('Buy'));
    expect(within(product2).getByText('In stock: 199')).toBeVisible();
    await userEvent.click(within(product2).getByText('Buy'));
    expect(within(product2).getByText('In stock: 198')).toBeVisible();

    await userEvent.click(screen.getByAltText('cart'));
    await screen.findByText('Shopping Cart');
    const modal = screen.getByTestId('modal-root');
    expect(within(modal).getByText('Broccoli')).toBeVisible();
    expect(within(modal).getByText('Cabbage')).toBeVisible();
    expect(within(modal).getByText('Total: 273,77 â‚¬')).toBeVisible();
    await userEvent.click(within(modal).getByText('Place order'));

    await screen.findByText('Order placed, thank you!');
    expect(fetchMock).toHaveBeenCalledTimes(2);
    expect(fetchMock.mock.calls[1]).toMatchSnapshot();
  });
});
