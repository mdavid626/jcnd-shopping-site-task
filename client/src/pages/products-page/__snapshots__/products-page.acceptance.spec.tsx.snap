// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Acceptance] products-page should be able to buy product 1`] = `
Array [
  "https://api.github.com/graphql",
  Object {
    "body": "{\\"operationName\\":\\"PlaceOrder\\",\\"variables\\":{\\"items\\":[{\\"productId\\":\\"63ea6485732ff5cf8b364431\\",\\"amount\\":1},{\\"productId\\":\\"63ea6485732ff5cf8b364434\\",\\"amount\\":2}]},\\"query\\":\\"mutation PlaceOrder($items: [OrderItem!]!) {\\\\n  placeOrder(items: $items)\\\\n}\\"}",
    "headers": Object {
      "accept": "*/*",
      "content-type": "application/json",
    },
    "method": "POST",
    "signal": AbortSignal {},
  },
]
`;
